### PR TITLE
refactor: dmacros.d to strings

### PR DESCRIPTION
**Pull request by WalterBright**
By using D strings instead of C strings, the complexity is reduced somewhat.

### CODE REVIEW

**Comment by mathias-lang-sociomantic** [comment]
I'm all for this change, but does LDC / GDC need to access it ?
If not, we should drop the `macro.h` header file. In any case, you'd need to update it as part of this P.R.

File: src/dmacro.d [hunk 1]
```diff
@@ -20,27 +20,23 @@ struct Macro
...
-    extern (C++) Macro* search(const(char)* name, size_t namelen)
+    Macro* search(const(char)[] name)
     {
         Macro* table;
-        //printf("Macro::search(%.*s)\n", namelen, name);
+        //printf("Macro::search(%.*s)\n", name.length, name.ptr);
         for (table = &this; table; table = table.next)
         {
-            if (table.namelen == namelen && memcmp(table.name, name, namelen) == 0)
+            if (table.name.length == name.length && memcmp(table.name.ptr, name.ptr, name.length) == 0)
```
Comment (ZombineDev) [inline src/dmacro.d]: `if (table.name == name`

File: src/dmacro.d [hunk 2]
```diff
@@ -50,21 +46,20 @@ private:
...
+    static Macro* define(Macro** ptable, const(char)[] name, const(char)[] text)
     {
-        //printf("Macro::define('%.*s' = '%.*s')\n", namelen, name, textlen, text);
+        //printf("Macro::define('%.*s' = '%.*s')\n", name.length, name.ptr, text.length, text.ptr);
         Macro* table;
         //assert(ptable);
         for (table = *ptable; table; table = table.next)
         {
-            if (table.namelen == namelen && memcmp(table.name, name, namelen) == 0)
+            if (table.name.length == name.length && memcmp(table.name.ptr, name.ptr, name.length) == 0)
```
Comment (ZombineDev) [inline src/dmacro.d]: `table.name == name`

File: src/dmacro.d [hunk 3]
```diff
@@ -251,18 +246,18 @@ public:
...
                             // Insert replacement text
-                            buf.spread(v + 1, 2 + m.textlen + 2);
+                            buf.spread(v + 1, 2 + m.text.length + 2);
                             buf.data[v + 1] = 0xFF;
                             buf.data[v + 2] = '{';
-                            memcpy(buf.data + v + 3, m.text, m.textlen);
-                            buf.data[v + 3 + m.textlen] = 0xFF;
-                            buf.data[v + 3 + m.textlen + 1] = '}';
-                            end += 2 + m.textlen + 2;
+                            memcpy(buf.data + v + 3, m.text.ptr, m.text.length);
```
Comment (ZombineDev) [inline src/dmacro.d]: `buf.data[v + 3 .. m.text.length] = m.text[]`

File: src/doc.d [hunk 4]
```diff
@@ -501,25 +501,25 @@ extern (C++) void gendocfile(Module m)
...
     {
         const(char)* p = m.toPrettyChars();
-        Macro.define(&m.macrotable, cast(char*)"TITLE", 5, cast(char*)p, strlen(p));
+        Macro.define(&m.macrotable, "TITLE", p[0 .. strlen(p)]);
     }
     // Set time macros
     {
         time_t t;
         time(&t);
         char* p = ctime(&t);
```
Comment (ZombineDev) [inline src/doc.d]: Next step - use `core.time.MonoTime.currTime.toString`

File: src/doc.d [hunk 5]
```diff
@@ -501,25 +501,25 @@ extern (C++) void gendocfile(Module m)
     // Set the title to be the name of the module
     {
         const(char)* p = m.toPrettyChars();
```
Comment (ZombineDev) [inline src/doc.d]: Next step - add alternative method that returns a slice

**Comment by ZombineDev** [comment]
LGTM too, minus @mathias-lang-sociomantic's comment and my tiny improvement ideas.

File: src/doc.d (hunk 4 above)
Comment (JackStouffer) [inline src/doc.d]: That allocates on the GC.

**Comment by WalterBright** [comment]
I just removed macro.h. Nobody used it anyway.

File: src/doc.d (hunk 4 above)
Comment (WalterBright) [inline src/doc.d]: Yes, it should stay the same.

File: src/dmacro.d [hunk 6]
```diff
@@ -251,18 +246,18 @@ public:
...
                             // Insert replacement text
-                            buf.spread(v + 1, 2 + m.textlen + 2);
+                            buf.spread(v + 1, 2 + m.text.length + 2);
                             buf.data[v + 1] = 0xFF;
                             buf.data[v + 2] = '{';
-                            memcpy(buf.data + v + 3, m.text, m.textlen);
-                            buf.data[v + 3 + m.textlen] = 0xFF;
-                            buf.data[v + 3 + m.textlen + 1] = '}';
-                            end += 2 + m.textlen + 2;
+                            buf.data[v + 3 .. v + 3 + m.text.length] = cast(ubyte[])m.text[];
```
Comment (ZombineDev) [inline src/dmacro.d]: Yeah what I meant. Sorry for the wrong suggestion before, I was on the phone which is not the best place write code :D

**Comment by ZombineDev** [comment]
@WalterBright thanks for addressing my suggestions! All LGTM.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Pull request merged**